[2단계 - 페이먼츠 모듈] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
안녕하세요, 마르코! 니야입니다 😊
주말 잘 보내셨는지요 ㅎ..ㅎ
2단계 미션 제출하겠습니다! 리뷰 잘 부탁드려요 🙇🏻‍♀️

# 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

## 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

## 2단계 - Refactoring
- 모듈화된 컴포넌트의 재사용성 및 확장성을 경험할 수 있다.
- 요구사항 변경에 따른 컴포넌트 리팩터링 및 개선을 할 수 있다.
- 선택) 실제 프로젝트에서의 컴포넌트 통합 및 활용할 수 있다.


# 🕵️ 셀프 리뷰(Self-Review)

## 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - `components` npm 배포 링크: [@sooyeoniya/components](https://www.npmjs.com/package/@sooyeoniya/components)
  - `hooks` npm 배포 링크: [@sooyeoniya/hooks](https://www.npmjs.com/package/@sooyeoniya/hooks)
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: [Storybook: Modal Component](https://6812fc18153d0f93012d5a97-uidzhzjyep.chromatic.com/)

## 리뷰 요청 & 논의하고 싶은 내용

이번 미션에서는 추상화의 기준과 컴포넌트의 존재 이유에 대해 고민을 많이 했던 것 같아요. 이와 관련하여 아래 내용들을 작성해봤습니다 :)

## 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점

### 💡 ButtonGroup 컴포넌트의 존재 의미

이번 미션을 하면서, ButtonGroup 컴포넌트의 존재 의미에 대해 고민했습니다. 

2단계 미션을 진행하면서 모달 내부에서 사용하는 Input과 Button 공통 컴포넌트를 구현했는데요, 이 과정에서 두 개의 버튼을 묶어서 스타일을 정의해 제공하는 ButtonGroup이라는 컴포넌트를 제공해주었습니다. https://github.com/woowacourse/react-modules/commit/4f4b4aa7350be5b40a8bb746ab00019d22ee2193

기존에는 ButtonGroupWrapper를 통해 버튼 묶음에 기본적인 레이아웃 스타일(display: flex, gap: 10px, width: 100%)을 적용했지만,
스토리북 구현하는 과정에서 사용자 입장에서 자유도가 부족하다는 점이 불편하게 느껴졌습니다.
그래서 아예 이런 스타일을 없애고 사용자가 자유롭게 스타일을 지정할 수 있도록 ButtonGroupWrapper 을 제거하였습니다. https://github.com/woowacourse/react-modules/commit/169cc31bdde9e74cc028bf5f9ef662bed1aab0a6

하지만, ButtonGroupWrapper를 제거했더니, ButtonGroup 자체가 단순히 Button 두 개를 감싸는 역할밖에 하지 않게 되어서 컴포넌트를 사용할만한 명확한 이점이 존재하지 않았고, 결국 해당 ButtonGroup 컴포넌트를 제거하기로 결정했습니다. 공통 컴포넌트를 단순히 묶는 것 이상의 역할(예를 들면, 스타일, 재사용성, 의도 등)이 명확해야만 해당 컴포넌트를 제공하는 의미가 있다고 생각했기 때문입니다. https://github.com/woowacourse/react-modules/commit/abdc3e5eb27b5ead666d92130c393816fc1e3920

추가적으로 ButtonGroup 제거한 이유는,
👉 ActionButtons(모달 상태를 전달받고 있는 모달에 특화된 한 쌍의 확인/취소 버튼 그룹) 같은 명확한 역할이 있는 컴포넌트가 이미 존재하고,
👉 버튼 묶음 스타일은 사용자가 Button 컴포넌트만 사용해도 직접 구성할 수 있기 때문에,
👉 이런 불필요한 추상화를 제거하는 것이 오히려 사용성과 유지보수 측면에서 낫다고 판단했습니다.

### 💡 컴포넌트 스타일 제공에 대한 고민
앞선 내용에서 ButtonGroupWrapper와 같은 스타일 래퍼를 제거했던 것처럼,
컴포넌트에서 어디까지 스타일을 제공해야 하고, 어디까지 사용자 확장을 허용해야 할지에 대해 고민했습니다.

결론적으로, 기본 UI(피그마 기반)와 직접 추가해준 light/dark 모드 디자인 시스템을 우선적으로 적용하여 일관된 디자인을 보장하도록 하고,
사용자가 필요 시 자유롭게 스타일을 커스터마이징할 수 있도록 유연성도 함께 제공하는 방향으로 정리했습니다.

예를 들어, Button 컴포넌트는 기본 스타일과 함께 variant를 통해 두 가지 색상 버전을 선택할 수 있게 되어 있으며,
추가 커스터마이징이 필요한 경우에는 ComponentProps<"button">을 상속받아 style, className 등의 속성을 prop으로 전달해 직접 수정할 수 있도록 구현했습니다.

즉, 디자인 시스템의 일관성과 사용자 확장성 사이의 균형을 고려한 컴포넌트 설계 방식을 적용했습니다.

### 💡 모달 구조 고민 (Compound Component 패턴)

2단계에서는 `AlertModal`, `ConfirmModal`, `PromptModal` 등 다양한 형태의 모달을 지원할 수 있도록 모달 라이브러리를 설계해야 했습니다. 이 과정에서 해당 모달들을 각각 별도의 컴포넌트로 직접 구현하고 `Modal`과 분리하여 `export`하는 방식은 **자유도와 확장성이 떨어진다** 고 판단했습니다.

따라서, 특정 모달 형태에 제한하지 않고 사용자가 **입력 필드나 버튼 등을 조합하여 다양한 모달 형태를 직접 구성할 수 있도록** `Compound Component` 패턴을 적용하여 구현해보았습니다. 아래와 같이 `Modal` 컴포넌트 하위에 필요한 서브 컴포넌트를 연결하여 사용하도록 설계했습니다.

```tsx
Modal.Button = Button;
Modal.Input = Input;

Modal.CancelButton = CancelButton;
Modal.ConfirmButton = ConfirmButton;
Modal.ActionButtons = ActionButtons;
```

현재는 총 5개의 컴포넌트를 `Modal`에 연결해두었고, 이 중 아래 3개의 버튼 컴포넌트는 모달에 특화된 기능을 수행하는 컴포넌트입니다!

* `CancelButton`: 닫기만 수행하는 취소 버튼
* `ConfirmButton`: 닫기 외에도 사용자가 정의한 작업 핸들러를 함께 실행하는 확인 버튼
* `ActionButtons`: `CancelButton`과 `ConfirmButton`을 함께 묶어주는 버튼 그룹

위 컴포넌트들은 모두 모달 내부에서 공유되는 **context를 기반으로 동작하고 있구요**, Compound Component 패턴의 핵심인 **상태 공유와 조합의 유연함** 을 반영해보았습니다!

비록 현재는 닫기 기능 외에 공유하는 상태가 있지는 않지만, 추후 모달 내에서 다양한 상태를 공유해야 할 경우를 대비한 **확장 가능하고 일관된 구조** 를 갖추는 데 의미가 있다고 생각했습니다.

## 2) 이번 리뷰를 통해 논의하고 싶은 부분

### 1. ActionButtons의 버튼 순서 변경 가능 여부
현재 ActionButtons 컴포넌트에서 버튼 순서가 고정되어 있어, 사용자가 원하는 순서대로 버튼을 배치할 수 없습니다. 예를 들어, 확인 버튼을 취소 버튼보다 먼저 배치하고 싶은 경우에도, 코드상에서 이를 쉽게 변경할 수 없는데요.
버튼 순서를 자유롭게 조정할 수 있도록 제공하는 것이 더 유연한 사용이 가능할 것 같습니다. 하지만, ActionButtons를 사용하지 않고 ConfirmButton, CancelButton을 직접 조합해서 사용하면 이미 순서 변경은 가능한 상황이기도 합니다. 그래서 오히려 ActionButtons 자체의 필요성에 대해서도 의문이 들기도 했습니다.
사용자가 이 순서를 커스터마이징할 수 있는 옵션을 제공하는 것이 좋을지, 아니면 기본적인 순서를 고정하고 커스터마이징을 최소화하는 방향이 좋을지 논의해보고 싶습니다. 

### 2. ActionButtons의 `onConfirm` props 필요성
현재 ActionButtons 컴포넌트는 onConfirm을 별도로 받아서 ConfirmButton에 전달하는 방식입니다. 하지만 confirmProps에서 onClick을 처리할 수 있기 때문에 onConfirm을 굳이 별도로 분리해서 제공할 필요가 있을지 고민되었습니다.
confirmProps 하나만으로 충분히 확인 버튼의 동작을 처리할 수 있다면, onConfirm을 제거하고 confirmProps로 모든 동작을 처리하도록 변경하는 것이 더 깔끔한 설계가 아닐까 생각이 듭니다. 하지만, 지금은 역할을 명확히 하기 위해서 명시적으로 onConfirm을 분리해둔 상태이구요, 이 방식이 정말 필요한지 마르코의 의견도 들어보고 싶습니다!

### 3. Button 컴포넌트의 제공 여부
ActionButtons, CancelButton, ConfirmButton은 모달에 특화된 버튼이기 때문에 별도로 제공하는 것이 맞고, Input 컴포넌트도 입력 필드를 위한 단일 컴포넌트이기 때문에 사용자에게 제공해주는 방식이 이해가 됩니다, 근데, 일반적인 기본 Button 컴포넌트는 사실 크게 필요할지 고민이 됩니다.
Button 컴포넌트는 스타일만 다르게 적용된 기본적인 버튼에 불과하고, 내부적으로 context를 공유하지 않기 때문에, 굳이 이를 별도로 제공할 필요가 없지 않나 싶습니다. 왜냐하면 그냥 사용자가 button 컴포넌트를 하나 만드는 것과 다를 바가 없기 때문입니다. 따라서, 지금처럼 Button을 Modal에 연결하여 따로 사용자에게 제공하기보다는, 그냥 모달에 특화된 ActionButtons, CancelButton, ConfirmButton만 제공하는 것이 더 적합하다는 느낌이 듭니다. 이 부분에 대해서 마르코 의견을 받아보고 싶습니다.

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

wmakerjun committed:
docs: PR 템플릿 개선

wmakerjun committed:
docs: PR 템플릿 학습 목표 추가

sooyeoniya committed:
Merge branch 'woowacourse:main' into step2

sooyeoniya committed:
feat: 카드 브랜드 기능 확장 구현

- 카드 번호에 따라 브랜드를 식별하여 포맷팅된 값으로 반환
- visa, masterCard, diners, amex, unionPay
- useCardNumber 반환값 수정: 카드 번호, 카드 에러 상태, 카드 번호 상태 변경 핸들러 함수, 카드 브랜드, 포맷팅된 값

sooyeoniya committed:
refactor: 카드 번호 포맷팅 함수 리팩토링

- 포맷해주는 로직과 데이터 분리
-> 확장성 및 유지보수하기 좋음 (브랜드 추가하기 훨씬 쉬움)

sooyeoniya committed:
refactor: 카드 브랜드 식별 및 포맷팅 로직 관심사 분리

- 각각을 파일로 분리 -> 재사용성 높이기 및 관심사 분리
- 여러 파일에서 사용하는 카드 브랜드 타입 따로 types 폴더 내부로 분리

sooyeoniya committed:
refactor: 카드 번호 자릿수 범위 상수화

sooyeoniya committed:
test: 카드 식별 및 포맷팅 관련 RTL 테스트 구현

- useCardNumber 에서 변경된 부분에 대하여 기존 테스트 수정
-> cardNumber 타입, cardNumberValidation 타입, etc.

sooyeoniya committed:
feat: small, medium, large 크기 옵션을 prop으로 전달받아 모달 크기 조절

- 모바일 환경일 때 모달이 가로로 꽉 차는 UX 이슈를 예방하기 위해 가로 margin 20px 추가 (물론 position이 bottom 인 경우에는 margin 0)

sooyeoniya committed:
refactor: 유지보수 향상을 위한 모달 컴포넌트 분리

sooyeoniya committed:
refactor: 커스텀 훅 폴더 구조 수정

- lib 폴더 내부에 있는 폴더 외부로 분리

sooyeoniya committed:
chore: 커스텀 훅 버전 업데이트 0.0.8

sooyeoniya committed:
refactor: 모달 폴더 구조 변경

- lib 폴더 내부에 있는 폴더들 외부로 분리

sooyeoniya committed:
rename: 모달 스토리 테스트 버튼 이름 변경

- 테스트 열기 버튼 -> 모달 열기

sooyeoniya committed:
fix: 모달 타입 import 경로 수정

sooyeoniya committed:
fix: 모달 컴포넌트 라이브러리 빌드 시 타입 include 경로 수정

sooyeoniya committed:
refactor: 공통 컴포넌트 폴더 묶기

sooyeoniya committed:
feat: Button 컴포넌트 구현

sooyeoniya committed:
feat: Button Varient 구현

- primary: 검정색 버튼
- secondary: 흰색 버튼

sooyeoniya committed:
feat: ButtonGroup 컴포넌트 구현

sooyeoniya committed:
design: 모달 padding 수정

- 피그마에 나온대로 재정의

sooyeoniya committed:
feat: Input 컴포넌트 구현

sooyeoniya committed:
refactor: 버튼 컴포넌트 확장성을 고려하여 text -> children 으로 변경

sooyeoniya committed:
refactor: ButtonGroup 타입 수정

- button 컴포넌트 타입 상속

sooyeoniya committed:
design: Input width 오류 수정

- Modal 내부에 Input을 넣었을 때, 모달에서 설정된 Padding을 무시하고 그대로 튀어나오는 현상 있었음

sooyeoniya committed:
refactor: ModalContent 컴포넌트 분리 취소 및 emotion으로 컴포넌트 설정

- ModalContent 파일을 따로 분리하는 것이 의미 없다고 판단함
- Modal 자체에 Button, ButtonGroup, Input 컴포넌트를 연결하여 사용자가 자유롭게 사용할 수 있도록 함

sooyeoniya committed:
feat: 모달 상태관리를 위한 Context 구현

- 컴파운드 컴포넌트 패턴의 핵심은 내부 상태를 모두 공유하는 것

sooyeoniya committed:
feat: theme 관련 객체 타입 명확히 지정

- satisfies 연산자를 사용해 객체 리터럴에 대해 타입을 명시적으로 지정해줌

sooyeoniya committed:
feat: 모달 테마 및 닫기 기능 props-> context 로 변경

sooyeoniya committed:
feat: 모달에 특화된 버튼 컴포넌트 3개 구현

[CancelButton]
- 기본 취소 버튼 기능
- useModalContext에서 제공하는 onClose 함수를 onClick으로 적용
- 모달 전체가 Context API를 사용해서 같은 상태를 공유하기 위함(컴파운드 컴포넌트)

[ConfirmButton]
- 확인 버튼 기능
- useModalContext에서 제공하는 onClose 함수를 onClick으로 적용
- 추가로 사용자가 직접 정의한 onClick 메서드가 있는 경우 먼저 실행
- 기본 동작으로 모달 닫기 실행 (onClose)

[ActionButtons]
- CancelButton, ConfirmButton을 통합하여 ButtonGroup과 유사한 버튼 그룹 생성
- ButtonGroup과 다른 점은 ButtonGroup은 순수 UI 컴포넌트, ActionButtons는 모달에 종속적인 컴포넌트(내부적으로 Context 사용)

sooyeoniya committed:
refactor: ConfirmButton, CancelButton 의 children 기본값 설정

- 각각 '확인', '취소'로 설정해두기
- UX 향상을 위함 -> children에 아무것도 사용자가 넣지 않아도
- 각 컴포넌트의 이름에 맞게 '확인', '취소'라는 이름으로 버튼이 보여야 함

sooyeoniya committed:
feat: Alert Modal 스토리 구현

- 일반 컴포넌트 Button과 모달에 특화된 ConfirmButton을 사용하여 2가지 스토리 구현
- 사용자가 직접 스타일 지정하여 좀 더 자유롭게 사용할 수 있도록, 기본적으로 제공해주는 스타일만 지정
- 사용자가 버튼 스타일 바꾸고 싶으면 props로 style 지정해서 넣어주면 됨 (Button 속성 상속받고 있음)

sooyeoniya committed:
refactor: ButtonGroup, ActionButtons 컴포넌트 Wrapper 스타일 제거

- 스토리 테스트 만드는 과정에서 사용자 입장에서의 스타일 자유도가 떨어져서 제거하기로 함

sooyeoniya committed:
feat: Confirm Modal 스토리 구현

sooyeoniya committed:
feat: Prompt Modal 스토리 구현

sooyeoniya committed:
refactor: ButtonGroup 컴포넌트 제거

- ButtonGroupWrapper 를 없앴더니 ButtonGroup 컴포넌트가 단순히 Button 두 개를 한 번 감싸주는 느낌만 들음
- 스타일을 지정해서 묶어주지 않으니 ButtonGroup 자체도 필요 없다고 생각함
- 따라서, ButtonGroup을 제거하고, Button 두 개를 직접 사용해서 구현할 수 있도록 함
- 어차피 ActionButtons 에서 두 개 묶음 컴포넌트를 이미 제공하고 있어서 ButtonGroup의 의미가 크게 없다고 느낌

sooyeoniya committed:
design: Button heigth 줄이기

sooyeoniya committed:
chore: 스토리북 알림 메시지 수정

sooyeoniya committed:
test: 스토리북 관심사별로 파일 분리

sooyeoniya committed:
test: 커스텀 훅 테스트 경로 오류 수정

sooyeoniya committed:
feat: 모달 포커스 트랩 기능 구현

sooyeoniya committed:
feat: 모달 열렸을 때 스크롤 방지 기능 구현

- restorePosition이 false 인 경우 모달이 닫혔을 때 scroll이 맨 위로 올라감

sooyeoniya committed:
chore: 불필요한 파일 제거

- 라이브러리를 위한 프로젝트이므로 기본 웹 배포 페이지 관련 코드 제거

sooyeoniya committed:
docs: 각 라이브러리 README 수정

sooyeoniya committed:
chore: npm 버전 업그레이드(컴포넌트, 훅 둘다)

sooyeoniya commented:
@coderabbitai pause

sooyeoniya committed:
fix: varient -> variant 오타 수정

wonsss commented on components/src/components/ModalActions/ConfirmButton.tsx:
```diff
@@ -0,0 +1,23 @@
+  const { onClose } = useModalContext();
+
+  const handleConfirmClick = (e: React.MouseEvent<HTMLButtonElement>) => {
+    if (onClick) onClick(e);
```
`onClick?.(e);` 처럼 옵셔널 체이닝 연산자를 사용하여 간단히 작성할 수 있어요.

sooyeoniya replied:
수정했습니다! 😁 https://github.com/woowacourse/react-modules/pull/142/commits/edfb81bfc398aca14043a15555355158fd65099a

wonsss commented on components/src/components/ModalActions/ActionButtons.tsx:
```diff
@@ -0,0 +1,31 @@
+export interface ActionButtonsProps {
+  confirmText?: ReactNode;
+  cancelText?: ReactNode;
+  onConfirm?: (e: React.MouseEvent<HTMLButtonElement>) => void;
```
onConfirm  prop 은 옵셔널이 아니라 필수값이어야 하지 않을까요?

그리고 단순히 궁금한 점입니닷. 혹시 event를 인자로 받을 수 있도록 설계하신 이유가 무엇일까요?
예를 들어, 어떤 경우에 Modal의 ConfirmButton의 onConfirm 핸들러에서 버튼의 event를 받아서 사용할 수 있는지 예시가 궁금합니다.

sooyeoniya replied:
저도 onConfirm의 선택/필수 여부에 대해 고민이 있었습니다. 실제로 ConfirmButton 컴포넌트에서도 onClick 타입에 undefined를 허용하고 있어, 기본적으로 옵셔널로 설정해두었습니다. 사용자 입장에서도 onConfirm이 필수가 아닌 쪽이 더 유연하다고 판단했어요.

그 이유는, 모달을 단순히 정보 표시 용도로만 사용하는 시나리오도 있기 때문입니다. 이런 경우에는 확인 버튼에 별도의 동작이 필요 없을 수 있어요. 그래서 확인 버튼은 기본적으로 모달을 닫는 기능만 제공하고, 필요한 경우에만 onConfirm을 통해 추가 동작을 넣을 수 있도록 했습니다.

만약 onConfirm을 필수로 만들게 되면, 단순히 모달을 닫기만 하는 상황에서도 불필요하게 빈 핸들러 함수(`() => {}`)를 전달해야 하는 문제가 생기기 때문에, 지금처럼 옵셔널로 두는 것이 더 실용적이라고 생각합니다.

---

`e: React.MouseEvent<HTMLButtonElement>`는 브라우저의 기본 동작을 막기 위해 `e.preventDefault()`를 사용하거나, `e.stopPropagation()`으로 이벤트 전파를 제어해야 할 상황에 대비해 이벤트 객체를 넘길 수 있도록 해둔 것이었습니다!

그런데 마르코가 던진 질문을 다시 생각해보니, **실제로 ‘모달’ 컴포넌트에서 이벤트 객체가 필요한 경우는 드물다는 결론에 도달했습니다.**

대부분의 모달 확인 버튼은 단순히 데이터를 저장하거나 상태를 업데이트하는 등의 **로직 실행용 콜백**을 호출하는 수준이기 때문에 이벤트 객체를 직접 다룰 필요가 없습니다. 서버로 요청을 보내는 경우에도 이벤트 객체를 참조할 일은 거의 없구요.

물론 이벤트 객체를 넘기면 이벤트 전파를 막거나 기본 동작을 제어하거나, 이벤트의 메타데이터를 확인하는 등의 작업이 가능하긴 합니다. 하지만 **‘모달’이라는 UI 맥락에서는 이런 기능이 실제로 거의 사용되지 않기 때문에**, 사용성 측면에서는 불필요하다고 판단했습니다.

다만 한 가지 고려해야 할 점은, ConfirmButton 컴포넌트에서 이벤트 객체를 제거하면 handleConfirmClick의 타입이 `() => void`로 추론되어, Button 컴포넌트가 기대하는 onClick 타입인 `(e: React.MouseEvent<HTMLButtonElement>) => void`와 일치하지 않게 된다는 것입니다. 그 이유가 onClick은 `React.MouseEventHandler<HTMLButtonElement>` 타입으로 정의되어 있어 이벤트 객체를 매개변수로 받도록 되어 있기 때문입니다.

결론적으로, 이벤트 객체를 실제로 사용하지 않더라도 **타입 시스템의 일관성과 하위 컴포넌트와의 호환성 유지**를 위해 현재 구조대로 이벤트 객체를 넘겨주는 방식으로 유지하기로 결정했습니다!

좋은 질문 덕분에 다시 한 번 고민해보며 더 나은 방향으로 정리할 수 있었습니다 :)
이벤트 관련 내용을 공부하다 보니, 정말 파도처럼 끝없이 새로운 개념들이 계속 나오네요 ㅎㅎ
덕분에 더 깊이 있게 공부해 볼 주제를 발견할 수 있었어요. 감사합니다! 😁

wonseokjang-29cm replied:
깊이 있는 답변 감사합니다! 말씀하신 것처럼 현재 구조 유지하는 게 적절할 것 같네요!

wonsss replied:
깊이 있는 답변 감사합니다! 말씀하신 것처럼 현재 구조 유지하는 게 적절할 것 같네요!

wonsss commented on components/src/components/ModalActions/ActionButtons.tsx:
```diff
@@ -0,0 +1,31 @@
+
+export interface ActionButtonsProps {
+  confirmText?: ReactNode;
+  cancelText?: ReactNode;
```
ReactNode는 버튼명에게는 타입이 넓습니다.
버튼명에는 문자열만 받도록 타입을 좁혀두는 게 안정적일 것 같습니다.

sooyeoniya replied:
CancelButton과 ConfirmButton에서는 내부에서 children으로 값을 받아 사용하고 있어서, 자연스럽게 ReactNode 타입으로 정의했던 것 같네요. ActionButtons에서는 confirmText와 cancelText가 버튼 텍스트 역할을 하고 있기 때문에, string 타입으로 수정했습니다 ㅎㅎ https://github.com/woowacourse/react-modules/pull/142/commits/ca3ce5abae9600fc01a7fcc3a764d9ab0bf2507b

다만, 확장성을 고려하면 향후 아이콘이나 다른 요소가 함께 들어갈 가능성도 있어 children 형태로 받는 방식도 유효하다고 생각했어요. 하지만 현재는 텍스트로만 고정해서 사용하는 방식이 예상 가능한 사용 패턴을 유지하고 불필요한 복잡도를 줄이는 데 도움이 된다고 판단해 string 타입으로 제한했습니다!

wonsss commented on components/src/components/Common/Button/Button.tsx:
```diff
@@ -0,0 +1,21 @@
+import { ButtonVariant } from "../../../constants/theme";
+import { ButtonWrapper } from "./Button.styles";
+
+export interface ButtonProps extends ComponentProps<"button"> {
+  variant?: ButtonVariant;
+}
```
variant와 ComponentProps<"button"> 까지 훌륭하네요. 👍

sooyeoniya replied:
감사합니다 😄

wonsss commented on components/src/constants/theme.ts:
```diff
@@ -0,0 +1,46 @@
+} as const;
+
+export type ThemeMode = "light" | "dark";
+export type ButtonVariant = "primary" | "secondary";
```
디자인시스템에서 시각적 위계를 나타내는 개념적 용어로 primary, secondary 를 잘 사용해주셨네요.
참고로 이외에도 primary, secondary, tertiary, ghost, destructive 등으로 컴포넌트의 우선순위와 성격을 구분합니다. 
이미 잘 활용해주신 것으로 봐서, 관련 내용을 잘 숙지하고 계신 것 같네요.
참고로 디자인시스템 링크 남겨둡니다.

https://carbondesignsystem.com/components/button/usage/#variants
https://m3.material.io/styles/color/roles

sooyeoniya replied:
공유해주신 링크들 참고해서 컴포넌트 우선순위와 성격을 더 명확하게 구분해보겠습니다!
앞으로도 일관된 디자인 시스템 구축에 많이 도움이 될 것 같아요 :)
좋은 말씀과 자료 감사합니다 🙇🏻‍♀️

wonsss commented on components/src/components/ModalContainer/ModalContainer.tsx:
```diff
@@ -0,0 +1,40 @@
+        $position={position}
+        $size={size}
+        $backgroundColor={currentTheme.background}
+        aria-labelledby={titleId}
+        aria-describedby={contentId}
+        onClick={stopPropagation}
+        role="dialog"
+        aria-modal="true"
```
aria 속성을 잘 적용해주셨네요. 👍 
각각의 의미와 역할을 설명해주실 수 있으실까요?

sooyeoniya replied:
지난번 [PR 리뷰](https://github.com/woowacourse/react-modules/pull/105#discussion_r2071691618)에 남겨주신 내용과 자료를 참고해서 구현했는데요,
각 의미와 역할에 대해서 아래 정리해보았습니다!

```tsx
<ModalWrapper ... role="dialog" aria-modal="true">
```

우선 모달 컴포넌트에 ARIA 속성을 적용하기 위해 ModalWrapper에 `role="dialog"`를 설정해주었습니다. 이 속성은 해당 요소가 대화 상자(dialog)임을 스크린 리더와 같은 보조 기술에 알려주는 역할을 합니다. 쉽게 말하자면, 이는 사용자가 현재 상호작용이 필요한 UI에 진입했음을 보조 기술이 인식할 수 있도록 도와주는 속성입니다.

`aria-modal="true"`의 경우에는 모달 뒤의 콘텐츠가 비활성화되어 있고, 현재 모달만이 상호작용 가능함을 나타내는 속성입니다. 즉, 스크린 리더가 모달 외부 콘텐츠를 무시할 수 있도록 해줍니다.

```tsx
const id = useId();
const titleId = `modal-title-${id}`;
const contentId = `modal-content-${id}`;
```

또한 ARIA 설정을 위해 제목 ID(`titleId`)와 내용 ID(`contentId`)는 Modal에서 useId 훅을 사용해 고유하게 생성하였습니다.

```tsx
// 제목 ID 지정
<Title ... id={titleId}>{titleText}</Title>

// 내용 ID 지정
<ModalContent id={contentId}>{children}</ModalContent>
```

이렇게 생성된 ID는 각각 제목과 내용 요소에 id로 지정해주고, `ModalWrapper`에 `aria-labelledby={titleId}`와 `aria-describedby={contentId}` 속성으로 연결해주었습니다.

```tsx
<ModalWrapper
  aria-labelledby={titleId}
  aria-describedby={contentId}
  ...
>
```

- `aria-labelledby={titleId}` 의 경우에는 모달의 제목을 지정합니다. 이 속성은 모달 제목 요소의 ID를 참조해서 스크린 리더가 모달을 식별할 수 있게 해줍니다.
- `aria-describedby={contentId}` 는 모달의 내용을 설명하는 요소를 지정해주는 건데요, 앞선 제목과 마찬가지로 ID를 지정하여 스크린 리더가 모달 내용을 식별하고 설명할 수 있게 해줍니다.

이렇게 하면 스크린 리더가 모달에 진입했을 때 제목과 내용을 정확하게 인식하고 읽어줄 수 있어, 접근성을 높일 수 있습니다.

닫기 버튼에도 접근성 속성을 추가해주었습니다.

```tsx
<CloseButton ... aria-label="close">
  <IoClose ... aria-hidden="true"/>
</CloseButton>
```

`aria-label="close"`는 시각적인 텍스트 없이 아이콘만 있는 닫기 버튼에 레이블을 제공하여, 스크린 리더가 이 버튼의 역할을 "close"라고 인식하고 읽어줄 수 있도록 도와줍니다.

또한 내부의 닫기 아이콘에는 `aria-hidden="true"`를 설정하여, 스크린 리더가 아이콘 자체를 읽지 않도록 했습니다.

```tsx
<ModalBackdrop ... role="presentation">
```

모달 백드롭에도 접근성 관련 역할을 부여하였는데요, 단순히 시각적인 표현만할 뿐 의미 없는 요소이기 때문에 해당 요소 자체는 스크린 리더가 무시하고, 자식 콘텐츠는 스크린 리더가 읽을 수 있도록 해당 역할을 `role="presentation"` 로 적용해주었습니다.

wonseokjang-29cm replied:
우와! 정말 잘 설명해주셨네요! 👍  꼼꼼하게 처리하시고 학습내용도 정리하시는 모습이 멋집니다

wonsss replied:
완벽한 답변이네요! 👍

wonsss commented on components/src/components/Modal/stories/Confirm.stories.tsx:
```diff
@@ -0,0 +1,122 @@
+    <>
+      <button type="button" onClick={openModal}>
+        모달 열기
+      </button>
+      <Modal
+        {...args}
+        isOpen={isOpen}
+        onClose={closeModal}
+        title={{ text: "카드를 삭제하시겠습니까?" }}
+      >
+        <p>삭제하면 복구하실 수 없습니다.</p>
+        <div
+          style={{
+            display: "flex",
+            justifyContent: "flex-end",
+            width: "100%",
+            gap: "10px",
+          }}
+        >
+          <Modal.Button
+            onClick={closeModal}
+            variant="secondary"
+            style={{ width: "90px" }}
+          >
+            취소
+          </Modal.Button>
+          <Modal.Button
+            onClick={handleSubmit}
+            variant="primary"
+            style={{ width: "90px" }}
+          >
+            확인
+          </Modal.Button>
+        </div>
+      </Modal>
```
이렇게 스토리북으로 해주신 것처럼 조합된 preset을 제공하는 AlertModal, ConfirmModal, PromptModal 컴포넌트도 함께 라이브러리에서 제공해주시면 편리할 것 같습니다.
사용자가 ConfirmModal 필요할 때마다, 조합을 직접 하다보면 실수도 날 수 있고 번거로울 것 같아요.

이처럼 compound component로 조합할 수 있는 유연성을 제공하면서, 자주 사용되는 케이스를 preset으로 제공하는 컴포넌트를 제공함으로써 편의성도 제공하면, 좋을 것 같아요.

sooyeoniya replied:
이 부분에 대해서 3가지 버전의 모달을 추가로 제공해줄까 고민을 많이 했는데, 제공해주는 편이 확실히 사용하기 편리할 것 같네요! 😊
3가지 프리셋 구현하고, 프리셋 컴포넌트를 사용한 스토리도 각각 추가해주었습니다! 
추가로, 라이브러리 사용을 위해 export 해주고, 프리셋 사용법에 대해서 리드미도 추가해두었어요.

- [Alert Modal Preset 구현](https://github.com/woowacourse/react-modules/pull/142/commits/0775b69a384cbd7875eb2fb72b145bd022cb17e5)
- [Prompt Modal Preset 구현](https://github.com/woowacourse/react-modules/pull/142/commits/cc02a8cfa4ba679260cc70c1ed9cb974a0b576b4) 
- [Confirm Modal Preset 구현](https://github.com/woowacourse/react-modules/pull/142/commits/ac1a799e83e62f67f8622bc6df21d7b6f5cb877c) 
- [라이브러리 사용을 위한 export](https://github.com/woowacourse/react-modules/pull/142/commits/99461f53e678e14d59b4bbb3fd2e2878bee0944b) 
- [README에 프리셋 사용법 정리](https://github.com/woowacourse/react-modules/pull/142/commits/f786b458352a746fe8839c8ccd198b05a681ccf5)
- [README에 프리셋 컴포넌트 props 설명 정리](https://github.com/woowacourse/react-modules/pull/142/commits/74701b0fdff60344a6b8ce1dbbb895646c5dab40)

wonseokjang-29cm replied:
👍  수고 많으셨습니다

wonsss replied:
수고 많으셨습니다! 👍

wonsss commented on components/src/components/Modal/stories/Interaction.stories.tsx:
```diff
@@ -0,0 +1,106 @@
+      });
+    });
+  },
+};
```
스토리북 상호작용 테스트 너무 좋습니다!! 👍 👍 👍 👍 

한 가지 추가 제안드리고 싶은 점은, Modal의 대표 기능에 대해서만 상호작용 테스트가 작성되어 있는데, 
ConfirmModal, PromptModal 등 케이스에서 confirm button/ cancel button 클릭 또는 prompt input 입력 등에 대한 상호작용 테스트도 각 케이스의 스토리북에 함께 있으면 테스트 케이스가 보완되어 좋을 것 같습니다.

sooyeoniya replied:
AlertModal, ConfirmModal, PromptModal 각 모달에 대한 상호작용 테스트 구현했습니다! 😃 https://github.com/woowacourse/react-modules/pull/142/commits/238d0db5b9b57c98c61215ef90cf84b3e7a9fc9b

wonsss commented on components/src/components/Modal/stories/Confirm.stories.tsx:
```diff
@@ -0,0 +1,122 @@
+  tags: ["autodocs"],
+};
+
+export const ConfirmModalWithButton = (args: ModalProps) => {
```
(중요도 낮음)
ConfirmModalWithActionButtons 외에 ConfirmModalWithButton 스토리북도 있는 이유가 궁금해요.
Modal.ActionButtons를 제공하시니까, ConfirmModalWithActionButtons 케이스가 다 포괄할 것 같아서요.

다른 Alert 등 스토리북들도 마찬가지 패턴으로 작성되어 있던데, 의도가 궁금했습니다.

sooyeoniya replied:
단순히 현재 Modal.Button을 제공하고 있다보니, 사용자에게도 Button을 사용해서 이런식으로 조합이 가능하다는 것을 스토리에서도 보여주고 싶었습니다.
['Button 컴포넌트의 제공 여부' 질문에 대한 마르코 답변](https://github.com/woowacourse/react-modules/pull/142#discussion_r2085037766) → 하지만 여기서 말씀하신것처럼 Modal.Button을 사용자에게 제공해주지 않는다면 이 스토리도 필요 없을 것 같습니다 :)

wonsss commented on components/src/components/ModalContainer/ModalContainer.tsx:
```diff
@@ -0,0 +1,40 @@
+import { ModalWrapper } from "./ModalContainer.styles";
+
+interface ModalContainerProps {
+  position: ModalPosition;
+  size: ModalSize;
```
ModalPosition과 ModalSize가 제한된 문자열 리터럴 타입으로 잘 정의되어 있네요 👍

sooyeoniya replied:
감사합니다! 🙇🏻‍♀️

wonsss commented on components/src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,107 @@
+  useBodyScrollLock({ isLocked: isOpen });
+  const { modalRef } = useModalFocusTrap(isOpen);
+
+  const id = useId();
```
useId() 훅은 어떤 역할을 하고, 이 경우 사용이 적절한 이유를 설명해주시면 좋을 것 같아요. 👍

sooyeoniya replied:
useId는 React 18에서 도입된 훅인데요, 접근성 속성에 주로 사용되는 고유한 ID를 생성해주는 역할을 하기 때문에 ARIA 속성을 추가하면서 이 훅을 사용하게 되었습니다.

실제 [공식문서](https://react.dev/reference/react/useId)의 첫 문장만 봐도 `“useId is a React Hook for generating unique IDs that can be passed to accessibility attributes.”` 라고 나와있고, 첫 번째 예시도 역시 접근성(`aria-describedby`) 속성을 설정하는 예제입니다.

단순히 문서에 그렇게 나와 있어서 사용하는 것이 아니라, 왜 이 상황에서 useId가 적절한지 그 이유를 좀 더 고민해봤습니다!

useId는 React가 고유 ID를 자동으로 관리해주기 때문에, 개발자가 수동으로 직접 ID를 관리하거나 충돌 방지를 위해 별도 로직을 작성할 필요가 없습니다.

또한, 가장 적합하다고 생각한 이유는 바로 **ID의 일관성 유지**라는 점인데요. 이 훅은 컴포넌트가 리렌더링 되어도 생성된 ID가 변하지 않고 유지됩니다. 이는 React 내부적으로 ID가 컴포넌트 인스턴스에 연결되어 있기 때문에, 해당 컴포넌트가 마운트되어 있는 동안에는 항상 동일한 ID가 유지된다고 합니다!

그래서 ARIA 속성과 연결된 요소들이 항상 안정적으로 연결되고, 동일한 모달을 여러 번 사용하더라도 각각 고유한 ID를 가질 수 있습니다.

기존에 주로 사용하던 방식인 Math.random() 같은 방식은 리렌더링할 때마다 새로운 ID를 생성하기 때문에 ID 불일치가 발생하고, 외부 라이브러리(uuid 등)를 사용하는 경우에는 외부 의존성을 갖게 되고, 마찬가지로 리렌더링 시 ID 변경될 위험이 존재합니다.

즉, useId 훅은 모달과 같은 컴포넌트에서 접근성 관련 ID를 관리할 때 최적의 선택이라고 생각합니다. 애초에 이 훅 자체가 접근성 속성에 ID를 안정적으로 연결하기 위해 도입된 훅이기도 하고, 모달이 열리고 닫히는 동안, 또는 내부 상태가 변경될 때도 ID가 안정적으로 유지되기 때문입니다!

wonseokjang-29cm replied:
완벽한 답변이네요! 👍

wonsss replied:
깊이 있게 잘 설명해주셨네요! 👍

wonsss commented on components/src/components/Modal/stories/Confirm.stories.tsx:
```diff
@@ -0,0 +1,122 @@
+      <button type="button" onClick={openModal}>
+        모달 열기
+      </button>
+      <Modal
+        {...args}
+        isOpen={isOpen}
+        onClose={closeModal}
+        title={{ text: "카드를 삭제하시겠습니까?" }}
+      >
+        <p>삭제하면 복구하실 수 없습니다.</p>
+        <div
+          style={{
+            display: "flex",
+            justifyContent: "flex-end",
+            width: "100%",
+            gap: "10px",
+          }}
+        >
+          <Modal.ActionButtons
+            onConfirm={handleSubmit}
+            confirmProps={{ style: { width: "90px" } }}
+            cancelProps={{ style: { width: "90px" } }}
+          />
+        </div>
+      </Modal>
```
> 1. ActionButtons의 버튼 순서 변경 가능 여부
현재 ActionButtons 컴포넌트에서 버튼 순서가 고정되어 있어, 사용자가 원하는 순서대로 버튼을 배치할 수 없습니다. 예를 들어, 확인 버튼을 취소 버튼보다 먼저 배치하고 싶은 경우에도, 코드상에서 이를 쉽게 변경할 수 없는데요.
버튼 순서를 자유롭게 조정할 수 있도록 제공하는 것이 더 유연한 사용이 가능할 것 같습니다. 하지만, ActionButtons를 사용하지 않고 ConfirmButton, CancelButton을 직접 조합해서 사용하면 이미 순서 변경은 가능한 상황이기도 합니다. 그래서 오히려 ActionButtons 자체의 필요성에 대해서도 의문이 들기도 했습니다.
사용자가 이 순서를 커스터마이징할 수 있는 옵션을 제공하는 것이 좋을지, 아니면 기본적인 순서를 고정하고 커스터마이징을 최소화하는 방향이 좋을지 논의해보고 싶습니다.

ActionButtons 내 CancelButton 과 ConfirmButton 의 순서를 조정할 수 있는 옵션을 제공할지에 대한 고민이시군요!
저는 순서를 조정하지 않아도 된다고 생각합니다. 
디자인시스템마다 의도에 따라 UX와 인터페이스를 다르게 설계할 수 있어서, 제 개인적인 의견임을 참고해주세요.
현재 CancelButton과 ConfirmButton 은 각각 시작적 위계상 secondary와 primary 입니다. 
primary와 seondary는 이름 그대로 다음과 같이 위계가 있어요. 
- primary: 가장 중요하고 주요한 행동 유도, 한 페이지에서 1개만 유지 권장 ex) 확인, 결제하기, 가입하기
- secondary: 보조적 선택이나 부가적 행동, ex) 취소, 또는 primary 옆 부가 기능으로서 사용

Primary 버튼이 우측에, Secondary 버튼이 좌측에 배치되는 이유는 주로 사용자의 시각 흐름과 결정 행동의 순서 때문입니다. 
서양권(좌→우 읽기 문화)의 시각 흐름은 왼쪽에서 오른쪽으로 이동합니다. 이 흐름에서 가장 마지막에 도달하는 위치(우측)가 가장 눈에 띄기 때문에, 중요한 결정 버튼(Primary)을 오른쪽에 배치합니다.
실제로 이 코멘트 작성 중에도 깃허브 코멘트의 버튼 그룹도 이와 같이 배치되어 있네요.
<img width="384" alt="image" src="https://github.com/user-attachments/assets/611df57a-7cc8-4f57-9f4b-ae37616334cb" />
[Google Material Design 가이드](https://m3.material.io/components/dialogs/guidelines#befd7f4d-1029-4957-b1b5-da13fc0bbf3c) 에서도 비슷한 내용을 설명하고 있으니, 참고해보세요.
<img width="1118" alt="image" src="https://github.com/user-attachments/assets/8b219e43-1058-4736-bcd4-05b3f6fae0da" />

wonsss replied:
> 3. Button 컴포넌트의 제공 여부
ActionButtons, CancelButton, ConfirmButton은 모달에 특화된 버튼이기 때문에 별도로 제공하는 것이 맞고, Input 컴포넌트도 입력 필드를 위한 단일 컴포넌트이기 때문에 사용자에게 제공해주는 방식이 이해가 됩니다, 근데, 일반적인 기본 Button 컴포넌트는 사실 크게 필요할지 고민이 됩니다.
Button 컴포넌트는 스타일만 다르게 적용된 기본적인 버튼에 불과하고, 내부적으로 context를 공유하지 않기 때문에, 굳이 이를 별도로 제공할 필요가 없지 않나 싶습니다. 왜냐하면 그냥 사용자가 button 컴포넌트를 하나 만드는 것과 다를 바가 없기 때문입니다. 따라서, 지금처럼 Button을 Modal에 연결하여 따로 사용자에게 제공하기보다는, 그냥 모달에 특화된 ActionButtons, CancelButton, ConfirmButton만 제공하는 것이 더 적합하다는 느낌이 듭니다. 이 부분에 대해서 마르코 의견을 받아보고 싶습니다.

네! 말씀대로 Button 컴포넌트는 Modal에서 제공하기보다는 별도로 자체적으로 Button을 디자인시스템에서 제공해주고, Modal에서는 이를 의존하는 방식이 적절합니다.
Modal.Button 처럼 인터페이스를 제공해줄 필요는 없을 것 같아요. 사실상 Modal.Button은 그냥 디자인시스템의 Button과 동일할 거라서요.

Modal 맥락에서 조립하기에 적절한 컴포넌트는 ActionButtons 이라고 생각해요.
이외에 굳이 CancelButton과 ConfirmButton도 제공할 필요가 없다고 생각해요.
디자인시스템은 인터페이스를 너무 많이 열어두기보다 UX 의도를 담아 적절한 제약을 걸어두는 것도 필요합니다. CancelButton과 ConfirmButton을 따로 제공하면, 사용자가 이 둘의 위치를 바꾼다거나 우측 정렬이 아니라 좌측 정렬을 한다거나 의도되지 않은 사용성을 만들 가능성이 있을 것 같습니다.

이보다는 레이아웃과 여러 가능성을 제약한 ActionButtons를 제공하고, ActionButtons는 single action / two actions 여부를 제어할 수 있도록 옵션을 제공하는 것도 검토해볼 수 있을 것 같아요. 
현재 ActionButtons 를 사용한다는 것은 무조건 두 버튼 모두 노출되는 것으로 개발되어 있어서요.

<img width="787" alt="image" src="https://github.com/user-attachments/assets/e36ded57-7912-4f29-af37-35c1ad7837c6" />

sooyeoniya replied:
> ActionButtons 내 CancelButton 과 ConfirmButton 의 순서를 조정할 수 있는 옵션을 제공할지에 대한 고민이시군요! 저는 순서를 조정하지 않아도 된다고 생각합니다. 디자인시스템마다 의도에 따라 UX와 인터페이스를 다르게 설계할 수 있어서, 제 개인적인 의견임을 참고해주세요. 현재 CancelButton과 ConfirmButton 은 각각 시작적 위계상 secondary와 primary 입니다. primary와 seondary는 이름 그대로 다음과 같이 위계가 있어요.
> 
> * primary: 가장 중요하고 주요한 행동 유도, 한 페이지에서 1개만 유지 권장 ex) 확인, 결제하기, 가입하기
> * secondary: 보조적 선택이나 부가적 행동, ex) 취소, 또는 primary 옆 부가 기능으로서 사용
> 
> Primary 버튼이 우측에, Secondary 버튼이 좌측에 배치되는 이유는 주로 사용자의 시각 흐름과 결정 행동의 순서 때문입니다. 서양권(좌→우 읽기 문화)의 시각 흐름은 왼쪽에서 오른쪽으로 이동합니다. 이 흐름에서 가장 마지막에 도달하는 위치(우측)가 가장 눈에 띄기 때문에, 중요한 결정 버튼(Primary)을 오른쪽에 배치합니다. 실제로 이 코멘트 작성 중에도 깃허브 코멘트의 버튼 그룹도 이와 같이 배치되어 있네요. 
 [Google Material Design 가이드](https://m3.material.io/components/dialogs/guidelines#befd7f4d-1029-4957-b1b5-da13fc0bbf3c) 에서도 비슷한 내용을 설명하고 있으니, 참고해보세요. 

디자인 시스템 관점에서 버튼의 시각적 위계와 배치 순서를 고정하는 이유를 예시로 들어서 구체적으로 설명해주셔서 감사합니다!

처음에는 단순히 커스터마이징 유연성 측면에서만 접근했는데, 말씀해주신 것처럼 시각 흐름이나 사용자 경험(UX)을 고려한 배치 방식이 실제로 명확한 목적과 일관성이 있다는 점에 대해 처음 알았어요. 우리나라도 확실히 좌→우 읽기 문화를 가지고 있다보니 primary 같은 중요한 결정 버튼은 오른쪽 edge에 배치하는 것이 훨씬 자연스러울 것 같네요!

공유해주신 Google Material Design 가이드도 자세히 보면서, 앞으로 버튼 배치 뿐만 아니라 컴포넌트 설계 측면에서 더 깊이 있는 고민 해볼 수 있을 것 같습니다.

정성스러운 답변 정말 감사합니다!! 😄

sooyeoniya replied:
> 네! 말씀대로 Button 컴포넌트는 Modal에서 제공하기보다는 별도로 자체적으로 Button을 디자인시스템에서 제공해주고, Modal에서는 이를 의존하는 방식이 적절합니다.
Modal.Button 처럼 인터페이스를 제공해줄 필요는 없을 것 같아요. 사실상 Modal.Button은 그냥 디자인시스템의 Button과 동일할 거라서요.
> Modal 맥락에서 조립하기에 적절한 컴포넌트는 ActionButtons 이라고 생각해요.
이외에 굳이 CancelButton과 ConfirmButton도 제공할 필요가 없다고 생각해요.
디자인시스템은 인터페이스를 너무 많이 열어두기보다 UX 의도를 담아 적절한 제약을 걸어두는 것도 필요합니다. CancelButton과 ConfirmButton을 따로 제공하면, 사용자가 이 둘의 위치를 바꾼다거나 우측 정렬이 아니라 좌측 정렬을 한다거나 의도되지 않은 사용성을 만들 가능성이 있을 것 같습니다.
> 이보다는 레이아웃과 여러 가능성을 제약한 ActionButtons를 제공하고, ActionButtons는 single action / two actions 여부를 제어할 수 있도록 옵션을 제공하는 것도 검토해볼 수 있을 것 같아요.
현재 ActionButtons 를 사용한다는 것은 무조건 두 버튼 모두 노출되는 것으로 개발되어 있어서요.

처음에는 ModalActions 폴더 내부에 있는 세 가지 컴포넌트(ActionButtons, ConfirmButton, CancelButton)를 모두 제공해서, 사용자가 원하는 방식대로 조합할 수 있도록 구성했었는데요.

마르코의 의견을 들어보니 ConfirmButton과 CancelButton은 ActionButtons와는 성격이 조금 다르다는 생각이 들었습니다.
ActionButtons는 말 그대로 사용자 액션을 유도하는 컴포넌트이고, Confirm/Cancel은 그 안에 조합된 구성 요소에 가깝기 때문이죠.

그래서 말씀하신 대로, ConfirmButton과 CancelButton을 따로 노출하기보다는, ActionButtons 하나만 제공하면서 single action / two actions 여부만 제어할 수 있도록 하는 게 훨씬 더 디자인 시스템적인 설계라는 생각이 들었습니다.

디자인 시스템은 무한한 커스터마이징을 허용하기보다는, 일정 수준의 제약을 통해 일관된 UX를 유도하는 것이 중요하다는 점에서 더 적절한 방향 같아요!

또한 CancelButton은 시각적 위계상 secondary 액션이기 때문에, single action일 경우에는 primary 액션(ConfirmButton)만 허용하고, two actions일 때에만 secondary를 함께 노출하는 구조가 훨씬 안정적이고 명확한 설계라고 생각했습니다.

말씀하신 것처럼 Button 자체는 Modal과는 별개의 범용적인 컴포넌트이므로, Modal에서 직접 제공하기보다는 별도로 제공하고, Modal에서는 ActionButtons만 의존하도록 수정해보았습니다! Button과 마찬가지로 Input도 별도로 제공할 수 있도록 수정했습니다.

- [컴포넌트 제공 관련 리팩토링](https://github.com/woowacourse/react-modules/pull/142/commits/8119333ae4295841674547455cb56c09a55ef1c9)
- [관련 리드미 수정](https://github.com/woowacourse/react-modules/pull/142/commits/2b3e17b6f77d1713e9ab34ae95be3d9c0dc19285)

wonsss replied:
👍 👍 👍

wonsss commented on components/src/components/ModalActions/ActionButtons.tsx:
```diff
@@ -0,0 +1,31 @@
+  return (
+    <>
+      <CancelButton {...cancelProps}>{cancelText}</CancelButton>
+      <ConfirmButton onClick={onConfirm} {...confirmProps}>
```
> 2. ActionButtons의 onConfirm props 필요성
현재 ActionButtons 컴포넌트는 onConfirm을 별도로 받아서 ConfirmButton에 전달하는 방식입니다. 하지만 confirmProps에서 onClick을 처리할 수 있기 때문에 onConfirm을 굳이 별도로 분리해서 제공할 필요가 있을지 고민되었습니다.
confirmProps 하나만으로 충분히 확인 버튼의 동작을 처리할 수 있다면, onConfirm을 제거하고 confirmProps로 모든 동작을 처리하도록 변경하는 것이 더 깔끔한 설계가 아닐까 생각이 듭니다. 하지만, 지금은 역할을 명확히 하기 위해서 명시적으로 onConfirm을 분리해둔 상태이구요, 이 방식이 정말 필요한지 마르코의 의견도 들어보고 싶습니다!

onConfirm 유지하는 게 가독성 및 유지보수성에 좋을 것 같습니다.
사용자는 confirmProps 인자 객체에 onClick을 전달해야 하는지를 명시적으로 알기 어려울 수 있을 것 같아요.
onConfirm의 타입도 필수값으로 변경해주신다면, 사용자는 무조건 onConfirm을 입력하게 되므로 더 쉽게 사용할 수 있게 됩니다.

sooyeoniya replied:
저도 onConfirm을 명시적으로 유지하는 것이 훨씬 가독성이 좋을 것 같다는 생각을 했습니다!
하지만, onConnfirm 타입의 옵션/필수 여부에 대해서는 좀 더 고민을 해봤습니다.

앞서 ActionButtons 컴포넌트를 single action과 two actions 모두 지원하도록 수정했는데요, 이 관점에서 보면 onConfirm을 옵션 값으로 유지하는 것이 자연스럽다고 생각합니다.

이에 대해서는 [이전 답변](https://github.com/woowacourse/react-modules/pull/142#discussion_r2084908399)에서도 언급드렸는데, 예시를 하나 더 들어보겠습니다!

[Google Material Design 가이드 - Dialog Single Button 예시](https://m3.material.io/components/dialogs/guidelines#b2bee645-4c24-4704-8fa4-c1aa815361c7)를 보면, OK 버튼 하나만 있는 단일 액션 모달이 존재합니다. 제 프로젝트에서의 AlertModal과 유사한 구조인데요, 이 경우 추가적인 액션 없이 모달을 닫기만 하면 되는 상황입니다.

ActionButtons를 사용하는 경우, 버튼들은 기본적으로 모달 컨텍스트에서 제공되는 닫기 기능을 수행합니다. 따라서 onConfirm을 넘겨주지 않아도 버튼 클릭 시 모달은 자동으로 닫히며, 알림 목적의 단순한 모달이라면 별도의 동작을 지정할 필요도 없습니다.

이런 구조라면 오히려 onConfirm을 필수로 강제하는 것보다, 지금처럼 선택적으로 두는 편이 더 실용적이고 자연스러운 설계라고 생각했습니다!

wonsss replied:
말씀대로 onConfirm을 유지하지만, 옵셔널로 처리하는 것이 적절할 것 같습니다 👍

wonsss commented on hooks/src/hooks/useCardNumber.ts:
```diff
@@ -0,0 +1,59 @@
+  formatCardNumber: string[];
+}
+
+const useCardNumber = (): UseCardNumberReturn => {
```
이건 제안이라서 한번 고민해보시고, 수정은 nice to have입니다.

카드 규칙을 라이브러리 외부에서 주입받도록 설계하면 어떨까요?
라이브러리 사용자 입장에서는 라이브러리를 수정하기 어렵습니다. 그런데 카드 규칙을 사용자가 주입할 수 있도록 한다면, 새로운 카드사가 추가되거나 규칙이 변경되어도 유연하게 대응할 수 있습니다. 또한, 순수 함수와 도메인 규칙이 분리되어 관심사의 분리가 명확해질 수 있어요.

추가적으로 사용자가 커스텀 규칙을 따로 주입하지 않으면, fallback으로서 라이브러리에서 제공하는 기본 카드사 규칙을 사용하도록 만드는 것도 유연한 방식 중 하나일 수 있어서 참고해주세요.

sooyeoniya replied:
확실히 좋은 제안인 것 같습니다!

현재는 카드 규칙이 라이브러리 내부에 하드코딩되어 있어, 사용자 입장에서는 이 라이브러리에서 제공하는 규칙 외에는 적용할 방법이 없는 상태이긴 합니다. 말씀해주신대로 외부에서 카드 규칙을 주입받을 수 있는 형태로 설계한다면, 새로운 카드사가 추가되거나 규칙이 변경되었을 때도 유연하게 대응할 수 있을 것 같습니다!

또한, 기본 규칙은 라이브러리 내부에 정의해두되, 사용자가 별도로 커스텀 규칙을 주입하지 않았을 경우 fallback으로 사용하도록 구성하면 확장성과 실용성 모두 챙길 수 있겠네요! 관심사의 분리라는 측면에서도 의미 있는 방향인 것 같고요. 

좋은 인사이트 주셔서 감사합니다! 😊

wonsss commented on hooks/src/utils/cardFormatter.ts:
```diff
@@ -0,0 +1,27 @@
+import { CardBrand, CardBrandType } from "../types/cardBrand";
+
+const CARD_FORMAT_SEGMENTS: Record<CardBrand, number[]> = {
+  visa: [4, 4, 4, 4],
+  masterCard: [4, 4, 4, 4],
+  diners: [4, 6, 4],
+  amex: [4, 6, 5],
+  unionPay: [4, 4, 4, 4],
+};
```
이 상수와 cardIdentifier.ts 에 있는 아래 상수는 카드사별 규칙이라는 관심사가 동일하므로, 해당 룰들은 단일 구조로 통합하는 것이 유지보수에 적절할 것 같습니다. 

```ts
const CARD_NUMBER_MAX_LENGTH: Record<CardBrand, number> = {
  visa: 16,
  masterCard: 16,
  diners: 14,
  amex: 15,
  unionPay: 16,
};
```

통합한다면, 나중에 새 카드사가 추가되거나, 기존 카드사가 삭제되었을 때, 유지보수가 용이합니다.

sooyeoniya replied:
처음에는 단순히 사용하는 곳에서 관심사 분리만 잘하면 유지보수하기 쉬울 것이라 생각했는데, 전체적으로 봤을 때는, 만약 새로운 카드사가 추가되는 경우, 수정해야할 곳이 많아져서 오히려 더 번거로울 것 같네요.

`CARD_RULES`라는 객체 상수를 만들어, 각 카드 브랜드별로 `maxLength`(번호 길이), `segments`(포맷 형식), `validator`(식별 로직)를 선언적으로 정의했습니다. 이 객체를 `cardFormatter`와 `cardIdentifier`에서 공통으로 참조할 수 있도록 수정했습니다! https://github.com/woowacourse/react-modules/pull/142/commits/a9b0bc01eb1302221a42a01565ba308eba0edb95

wonsss commented on hooks/src/__test__/useCardNumber.test.ts:
```diff
@@ -1,136 +1,288 @@
+          cardNumber: "36222233334444",
+          expectedFormatCardNumber: ["3622", "223333", "4444"],
+        },
+        {
+          description:
+            "American Express 카드는 4자리-6자리-5자리 형식으로 포맷팅된다",
+          cardBrand: "amex",
+          cardNumber: "341222233334444",
+          expectedFormatCardNumber: ["3412", "222333", "34444"],
+        },
+        {
+          description: "UnionPay 카드는 4자리씩 4개 그룹으로 포맷팅된다",
+          cardBrand: "unionPay",
+          cardNumber: "6221261111222233",
+          expectedFormatCardNumber: ["6221", "2611", "1122", "2233"],
+        },
+      ];
+
+      it.each(formatTestCases)(
+        "$description",
+        ({ cardBrand, cardNumber, expectedFormatCardNumber }) => {
+          // given
+          // when
+          const { result } = renderHook(() => useCardNumber());
+
+          act(() => {
+            result.current.handleCardNumberChange(cardNumber);
+          });
+
+          // then
+          expect(result.current.cardNumber).toEqual(cardNumber);
+          expect(result.current.cardNumberValidation).toEqual(
+            defaultValidationValue
+          );
+          expect(result.current.cardBrand).toEqual(cardBrand);
+          expect(result.current.formatCardNumber).toEqual(
+            expectedFormatCardNumber
+          );
+        }
+      );
```
테스트가 꼼꼼히 잘 작성되어 있네요!

wonsss reviewed (changes requested):
안녕하세요! 니야 😄 
2단계 미션 수고 많으셨습니다.
코드가 아주 깔끔하고 가독성이 좋네요. 접근성도 꼼꼼히 잘 챙겨주셔서 너무 좋았습니다. 👍 
질문 주신 부분들은 관련 코드 맥락에 코멘트 달아두었습니다.
리뷰 반영 후, 재리뷰 요청주세요! 감사합니다 🙇

wonsss commented on components/src/components/Common/Button/Button.tsx:
이건 미션요구사항이 아닌 것 같아 참고만 해주셔도 되고, 여유가 되시면 해보셔도 좋습니다.

만약, 버튼 동작이 '비동기' 작업인 경우는 디자인시스템에서 어떻게 대응해야 할까요? 실제 사용 케이스에서는 버튼이 API에 post 요청을 보내는 작업일 가능성이 커요.
예를 들어, 비동기 작업 중에는 버튼의 문구가 '로딩 스피너'로 변경되고 추가 클릭되지 않도록 disabled 되어야 하고, 작업이 완료된 후에 모달이 닫히도록 해야 할 거예요.

sooyeoniya replied:
좋은 포인트 감사합니다! 😃
실제로 버튼 액션이 비동기 작업인 경우가 많은데, 지금까지는 그 흐름을 구체적으로 고려하진 않았던 것 같네요.

> 만약, 버튼 동작이 '비동기' 작업인 경우는 디자인시스템에서 어떻게 대응해야 할까요?

이에 대해 답변 드려보자면, 디자인 시스템 관점에서는 다음과 같은 대응이 필요할 것 같습니다.

- 버튼 클릭 시 로딩 상태를 시각적으로 보여주기 위해, 말씀하신 것처럼 스피너나 ‘처리 중’ 같은 텍스트를 버튼 안에 표시하는 방식이 추가되어야 할 것 같아요.
    - 이때 버튼에 isLoading과 같은 prop을 제공하여 비동기 함수에 대응할 수 있도록 지원해야할 것 같습니다.
- 말씀해주신 것처럼, 로딩 중에는 버튼을 비활성화하여 중복 요청을 막는 것이 UX 측면에서 더 나은 설계인 것 같습니다.
- 비동기 작업이 완료된 후 모달을 닫거나 알림을 표시하는 등의 후속 동작에 대한 일관된 UX 흐름도 함께 고려해야 할 것 같습니다.

마침 지금 진행 중인 "상품 목록" 미션에서도 상품을 추가하거나 삭제할 때 비동기 API 통신을 처리해야 하는 부분이 있는데요. 이 흐름을 잘 참고해서, 해당 미션에서 버튼과 로딩 상태의 UX를 잘 녹여보도록 하겠습니다! 버튼 설계할 때 단순 클릭 이상의 흐름을 염두에 둬야 한다는 걸 다시 한 번 인지하게 됐던 것 같아요, 말씀 정말 감사합니다! 😊

wonsss replied:
잘 이해해주셨네요. 훌륭합니다!! 
사용자의 인터넷 상황이 느린 경우, POST 요청 버튼 클릭한 후 응답이 바로 오지 않으면, 여러 번 클릭하게 되어, 불필요한 API 요청이 발생하거나 좋지 않은 UX가 발생할 가능성이 있는데, 이런 상황까지 미리 고려하여 개발하는 태도가 중요하다고 생각합니다.

sooyeoniya committed:
refactor: 옵셔널 체이닝으로 조건문 수정

sooyeoniya committed:
fix: ActionButtons의 버튼 텍스트 props 타입 수정

- ReactNode -> string 으로 타입 좁히기

sooyeoniya committed:
feat: Alert Modal Preset 구현

Modal 내부로 전달되는 속성 중 required 값만 구조 분해하여 명시적으로 작성
- isOpen, onClose

sooyeoniya committed:
feat: Prompt Modal Preset 구현

sooyeoniya committed:
feat: Confirm Modal Preset 구현

sooyeoniya committed:
feat: 프리셋 컴포넌트 export

sooyeoniya committed:
docs: 리드미에 프리셋 컴포넌트 사용법 추가

sooyeoniya committed:
docs: 프리셋 컴포넌트 props 설명 추가

sooyeoniya committed:
refactor: 카드 브랜드별 식별 및 포맷 규칙 리팩토링

- 객체로 만들어 선언적으로 리팩토링

sooyeoniya committed:
refactor: Modal에서 ActionButtons만 제공되도록 수정

- ActionButtons에서 확인 / 확인취소 버튼을 모두 사용할 수 있도록 수정함
- Modal에서 ConfirmButton, CancelButton을 내보내지 않도록 수정
- Button과 Input은 Modal 과는 별개로 제공할 수 있도록 수정

sooyeoniya committed:
docs: 컴포넌트 제공 관련 리팩토링에 따라 문서 수정

sooyeoniya committed:
test: Alert, Confirm, Prompt 모달 상호작용 테스트

sooyeoniya commented:
안녕하세요, 마르코! 니야입니다 :)
“페이먼츠 모듈” 미션 2단계 PR에 대해 리뷰해주신 내용들을 반영해보았습니다.

이번까지 총 두 번의 미션을 마르코와 함께하게 되었는데, 
매번 깊이 있는 리뷰 덕분에 코드에 대해 더 많이 고민하고, 제 생각을 구체화하면서 구현해 볼 수 있었던 것 같아요.

개인적으로, 꼼꼼하게 답변도 잘 해주시고 고민의 방향성도 잘 잡아주신 덕분에 가장 많이 배운 리뷰어가 마르코라고 생각합니다.
혹시 리뷰 반영 사항 중 더 이야기해보고 싶은 부분이 있으시거나, 보완이 필요한 부분이 있다면 편하게 말씀해주시면 감사하겠습니다!

마지막 미션에서도 또 함께할 수 있다면 정말 좋을 것 같습니다.
진심으로 감사드립니다 🙇🏻‍♀️

아래는 npm 및 storybook 배포 링크입니다!
- components npm 배포 링크: [@sooyeoniya/components](https://www.npmjs.com/package/@sooyeoniya/components)
- hooks npm 배포 링크: [@sooyeoniya/hooks](https://www.npmjs.com/package/@sooyeoniya/hooks)
- Storybook 링크: [Storybook: Modal Component](https://6812fc18153d0f93012d5a97-uckasnmuvx.chromatic.com/)

sooyeoniya committed:
chore: npm 버전 업

sooyeoniya committed:
chore: 컴포넌트 라이브러리 npm 버전 업 및 스토리북 배포 링크 수정

wonsss reviewed (approved):
안녕하세요! 니야 😄 
2단계 페이먼츠 미션 리뷰 반영하시느라 수고 많으셨습니다.
리뷰 드린 모든 코멘트들에 대해 깊게 고민해주시고 정성스러운 답변을 남겨주셔서 감사합니다!
덕분에 저도 니야로부터 많이 배우게 되었던 시간이었습니다.  🙇 
이번 미션에는 이미 리뷰를 모두 잘 반영해주셨고 미션 목표에도 충분해진 것으로 보여 마무리하겠습니다.
앞으로도 정말 잘하실 것 같네요!! 화이팅입니다~! 💪
